source/mesh_refinement/thermal_energy_density.cc

```diff
@@ -101,11 +101,36 @@ namespace aspect
                                       this->get_mpi_communicator());
       vec = vec_distributed;
 
-      DerivativeApproximation::approximate_gradient  (this->get_mapping(),
-                                                      this->get_dof_handler(),
-                                                      vec,
-                                                      indicators,
-                                                      this->introspection().component_indices.temperature);
+      try
+        {
+          DerivativeApproximation::approximate_gradient  (this->get_mapping(),
+                                                          this->get_dof_handler(),
+                                                          vec,
+                                                          indicators,
+                                                          this->introspection().component_indices.temperature);
+        }
+      catch (std::exception &exc)
+        {
+          std::cerr << std::endl << std::endl
+                    << "----------------------------------------------------"
+                    << std::endl;
+          std::cerr << "Exception on MPI process <"
+                    << Utilities::MPI::this_mpi_process(MPI_COMM_WORLD)
+                    << "> while running mesh refinement plugin <"
+                    << "thermal energy density"
+                    << ">: " << std::endl
+                    << "The mesh refinement plugin failed to compute a thermal energy density gradient. "
+                    << "This could be caused by having a constant temperature in the model. If this is "
+                    << "the case, please set an other mesh refinement strategy. DEAL.II error:"
+                    << exc.what() << std::endl
+                    << "Aborting!" << std::endl
+                    << "----------------------------------------------------"
+                    << std::endl;
+
+          // terminate the program!
+          MPI_Abort (MPI_COMM_WORLD, 1);
```

Comment:
I would prefer if we handled aborting the program in one central place. If you take a look at `source/initial_temperature/function.cc` (or any other of the similar classes), we just call `throw` there and let things further up deal with it.